File: app/src/main/java/co/tpcreative/supersafe/common/controller/SingletonScreenLock.java
package co.tpcreative.supersafe.common.controller;
import android.os.CountDownTimer;

import co.tpcreative.supersafe.common.util.Utils;
import co.tpcreative.supersafe.model.EmailToken;
import co.tpcreative.supersafe.model.EnumPinAction;
import co.tpcreative.supersafe.model.EnumStatus;

public class SingletonScreenLock {

    private static SingletonScreenLock instance ;
    private static final String TAG = SingletonScreenLock.class.getSimpleName();
    public static SingletonScreenLock getInstance(){
        synchronized (SingletonScreenLock.class){
            if (instance==null){
                instance = new SingletonScreenLock();
            }
            return instance;
        }
    }

    private CountDownTimer mCountDownTimer;
    private SingletonScreenLock.SingletonScreenLockListener ls;

    public void setListener(SingletonScreenLock.SingletonScreenLockListener ls){
        this.ls = ls;
    }

    public void onStartTimer(long value){
        Utils.Log(TAG,"onStartTimer");
        if (mCountDownTimer!=null){
            Utils.Log(TAG,"Running............");
            return;
        }
        Utils.Log(TAG,"Start");
        mCountDownTimer = new CountDownTimer(value, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                long secondsRemaining = millisUntilFinished / 1000;
                SingletonScreenLock.getInstance().onAttemptTimer(""+secondsRemaining);
            }
            @Override
            public void onFinish() {
                Utils.Log(TAG,"Finish :");
                mCountDownTimer= null;
                SingletonScreenLock.getInstance().onAttemptTimerFinish();
            }
        }.start();
    }

    public void onStop(){
        if (mCountDownTimer!=null){
            mCountDownTimer.cancel();
            mCountDownTimer.onFinish();
            mCountDownTimer = null;
        }
    }

    public void onAttemptTimer(final String seconds){
        if (ls!=null){
            ls.onAttemptTimer(seconds);
        }
    }

    public void onAttemptTimerFinish(){
        if (ls!=null){
            ls.onAttemptTimerFinish();
        }
    }

    public interface SingletonScreenLockListener{
        void onAttemptTimer(String seconds);
        void onAttemptTimerFinish();
    }

}

File: app/src/main/java/co/tpcreative/supersafe/common/controller/SingletonScreenLock.kt
package co.tpcreative.supersafe.common.controller
import android.os.CountDownTimer
import co.tpcreative.supersafe.common.util.Utils
class SingletonScreenLock {
    private var mCountDownTimer: CountDownTimer? = null
    private var ls: SingletonScreenLockListener? = null
    fun setListener(ls: SingletonScreenLockListener?) {
        this.ls = ls
    }

    fun onStartTimer(value: Long) {
        Utils.Log(TAG, "onStartTimer")
        if (mCountDownTimer != null) {
            Utils.Log(TAG, "Running............")
            return
        }
        Utils.Log(TAG, "Start")
        mCountDownTimer = object : CountDownTimer(value, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                val secondsRemaining = millisUntilFinished / 1000
                getInstance()?.onAttemptTimer("" + secondsRemaining)
            }

            override fun onFinish() {
                Utils.Log(TAG, "Finish :")
                mCountDownTimer = null
                getInstance()?.onAttemptTimerFinish()
            }
        }.start()
    }

    fun onStop() {
        if (mCountDownTimer != null) {
            mCountDownTimer?.cancel()
            mCountDownTimer?.onFinish()
            mCountDownTimer = null
        }
    }

    fun onAttemptTimer(seconds: String?) {
        if (ls != null) {
            ls?.onAttemptTimer(seconds)
        }
    }

    fun onAttemptTimerFinish() {
        if (ls != null) {
            ls?.onAttemptTimerFinish()
        }
    }

    interface SingletonScreenLockListener {
        open fun onAttemptTimer(seconds: String?)
        open fun onAttemptTimerFinish()
    }

    companion object {
        private var instance: SingletonScreenLock? = null
        private val TAG = SingletonScreenLock::class.java.simpleName
        fun getInstance(): SingletonScreenLock? {
            synchronized(SingletonScreenLock::class.java) {
                if (instance == null) {
                    instance = SingletonScreenLock()
                }
                return instance
            }
        }
    }
}